File: src/572_subtree_of_another_tree/SubtreeOfAnotherTree.java

/**
 * 572. Subtree of Another Tree
 * 
 * Given the roots of two binary trees root and subRoot, return true if there is a subtree 
 * of root with the same structure and node values of subRoot and false otherwise.
 * 
 * A subtree of a binary tree tree is a tree that consists of a node in tree and all of 
 * this node's descendants. The tree tree could also be considered as a subtree of itself.
 */

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    /**
     * Check if subRoot is a subtree of root.
     * 
     * @param root    Root of the main binary tree
     * @param subRoot Root of the potential subtree
     * @return true if subRoot is a subtree of root, false otherwise
     */
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null) {
            return false;
        }
        if (isSameTree(root, subRoot)) {
            return true;
        }

        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }

    private boolean isSameTree(TreeNode a, TreeNode b) {
        if (a == null || b == null) {
            return a == b;
        }

        if (a.val != b.val) {
            return false;
        }

        return isSameTree(a.left, b.left) && isSameTree(a.right, b.right);
    }
}

public class SubtreeOfAnotherTree {
    /**
     * Helper function to build a binary tree from an array (level-order).
     */
    private static TreeNode buildTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null;
        }

        TreeNode root = new TreeNode(values[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;

        while (!queue.isEmpty() && i < values.length) {
            TreeNode node = queue.poll();

            if (i < values.length && values[i] != null) {
                node.left = new TreeNode(values[i]);
                queue.offer(node.left);
            }
            i++;

            if (i < values.length && values[i] != null) {
                node.right = new TreeNode(values[i]);
                queue.offer(node.right);
            }
            i++;
        }

        return root;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test case 1: subRoot is subtree at left child
        System.out.println("Test 1: subRoot is subtree at left child");
        TreeNode root1 = buildTree(new Integer[] { 3, 4, 5, 1, 2 });
        TreeNode subRoot1 = buildTree(new Integer[] { 4, 1, 2 });
        boolean result1 = solution.isSubtree(root1, subRoot1);
        System.out.println("Result: " + result1 + ", Expected: true");
        assert result1 == true : "Test 1 failed";

        // Test case 2: Extra node in root's subtree
        System.out.println("\nTest 2: Extra node in root's subtree");
        TreeNode root2 = buildTree(new Integer[] { 3, 4, 5, 1, 2, null, null, null, null, 0 });
        TreeNode subRoot2 = buildTree(new Integer[] { 4, 1, 2 });
        boolean result2 = solution.isSubtree(root2, subRoot2);
        System.out.println("Result: " + result2 + ", Expected: false");
        assert result2 == false : "Test 2 failed";

        // Test case 3: subRoot equals entire root
        System.out.println("\nTest 3: subRoot equals entire root");
        TreeNode root3 = buildTree(new Integer[] { 1, 2, 3 });
        TreeNode subRoot3 = buildTree(new Integer[] { 1, 2, 3 });
        boolean result3 = solution.isSubtree(root3, subRoot3);
        System.out.println("Result: " + result3 + ", Expected: true");
        assert result3 == true : "Test 3 failed";

        // Test case 4: Single node trees, matching
        System.out.println("\nTest 4: Single node trees, matching");
        TreeNode root4 = buildTree(new Integer[] { 1 });
        TreeNode subRoot4 = buildTree(new Integer[] { 1 });
        boolean result4 = solution.isSubtree(root4, subRoot4);
        System.out.println("Result: " + result4 + ", Expected: true");
        assert result4 == true : "Test 4 failed";

        // Test case 5: Single node trees, not matching
        System.out.println("\nTest 5: Single node trees, not matching");
        TreeNode root5 = buildTree(new Integer[] { 1 });
        TreeNode subRoot5 = buildTree(new Integer[] { 2 });
        boolean result5 = solution.isSubtree(root5, subRoot5);
        System.out.println("Result: " + result5 + ", Expected: false");
        assert result5 == false : "Test 5 failed";

        // Test case 6: subRoot at right child
        System.out.println("\nTest 6: subRoot at right child");
        TreeNode root6 = buildTree(new Integer[] { 3, 4, 5, 1, 2 });
        TreeNode subRoot6 = buildTree(new Integer[] { 5 });
        boolean result6 = solution.isSubtree(root6, subRoot6);
        System.out.println("Result: " + result6 + ", Expected: true");
        assert result6 == true : "Test 6 failed";

        // Test case 7: subRoot is leaf node
        System.out.println("\nTest 7: subRoot is leaf node");
        TreeNode root7 = buildTree(new Integer[] { 3, 4, 5, 1, 2 });
        TreeNode subRoot7 = buildTree(new Integer[] { 2 });
        boolean result7 = solution.isSubtree(root7, subRoot7);
        System.out.println("Result: " + result7 + ", Expected: true");
        assert result7 == true : "Test 7 failed";

        // Test case 8: Similar structure but different values
        System.out.println("\nTest 8: Similar structure but different values");
        TreeNode root8 = buildTree(new Integer[] { 3, 4, 5, 1, 2 });
        TreeNode subRoot8 = buildTree(new Integer[] { 4, 1, 3 });
        boolean result8 = solution.isSubtree(root8, subRoot8);
        System.out.println("Result: " + result8 + ", Expected: false");
        assert result8 == false : "Test 8 failed";

        // Test case 9: subRoot larger than root
        System.out.println("\nTest 9: subRoot larger than root");
        TreeNode root9 = buildTree(new Integer[] { 1, 2 });
        TreeNode subRoot9 = buildTree(new Integer[] { 1, 2, 3 });
        boolean result9 = solution.isSubtree(root9, subRoot9);
        System.out.println("Result: " + result9 + ", Expected: false");
        assert result9 == false : "Test 9 failed";

        // Test case 10: Complex tree, subRoot deep inside
        System.out.println("\nTest 10: Complex tree, subRoot deep inside");
        TreeNode root10 = buildTree(new Integer[] { 1, 2, 3, 4, 5, 6, 7, 8, 9 });
        TreeNode subRoot10 = buildTree(new Integer[] { 4, 8, 9 });
        boolean result10 = solution.isSubtree(root10, subRoot10);
        System.out.println("Result: " + result10 + ", Expected: true");
        assert result10 == true : "Test 10 failed";

        // Test case 11: Partial match (substring-like scenario)
        System.out.println("\nTest 11: Value appears but not as subtree");
        TreeNode root11 = buildTree(new Integer[] { 12 });
        TreeNode subRoot11 = buildTree(new Integer[] { 2 });
        boolean result11 = solution.isSubtree(root11, subRoot11);
        System.out.println("Result: " + result11 + ", Expected: false");
        assert result11 == false : "Test 11 failed";

        System.out.println("\n✅ All tests passed!");
    }
}

File: src/572_subtree_of_another_tree/subtree_of_another_tree.py
"""
572. Subtree of Another Tree

Given the roots of two binary trees root and subRoot, return true if there is a subtree
of root with the same structure and node values of subRoot and false otherwise.

A subtree of a binary tree tree is a tree that consists of a node in tree and all of
this node's descendants. The tree tree could also be considered as a subtree of itself.
"""

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        """
        Check if subRoot is a subtree of root.

        Args:
            root: Root of the main binary tree
            subRoot: Root of the potential subtree

        Returns:
            True if subRoot is a subtree of root, False otherwise
        """

        if not root:
            return False

        def isSameTree(a, b):
            if not a or not b:
                return a == b

            if a.val != b.val:
                return False

            return isSameTree(a.left, b.left) and isSameTree(a.right, b.right)

        if isSameTree(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)


# Test cases
def build_tree(values):
    """Helper function to build a binary tree from a list (level-order)."""
    if not values:
        return None

    root = TreeNode(values[0])
    queue = [root]
    i = 1

    while queue and i < len(values):
        node = queue.pop(0)

        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1

    return root


def run_tests():
    solution = Solution()

    # Test case 1: subRoot is subtree at left child
    print("Test 1: subRoot is subtree at left child")
    root1 = build_tree([3, 4, 5, 1, 2])
    subRoot1 = build_tree([4, 1, 2])
    result1 = solution.isSubtree(root1, subRoot1)
    print(f"Result: {result1}, Expected: True")
    assert result1 == True, "Test 1 failed"

    # Test case 2: Extra node in root's subtree
    print("\nTest 2: Extra node in root's subtree")
    root2 = build_tree([3, 4, 5, 1, 2, None, None, None, None, 0])
    subRoot2 = build_tree([4, 1, 2])
    result2 = solution.isSubtree(root2, subRoot2)
    print(f"Result: {result2}, Expected: False")
    assert result2 == False, "Test 2 failed"

    # Test case 3: subRoot equals entire root
    print("\nTest 3: subRoot equals entire root")
    root3 = build_tree([1, 2, 3])
    subRoot3 = build_tree([1, 2, 3])
    result3 = solution.isSubtree(root3, subRoot3)
    print(f"Result: {result3}, Expected: True")
    assert result3 == True, "Test 3 failed"

    # Test case 4: Single node trees, matching
    print("\nTest 4: Single node trees, matching")
    root4 = build_tree([1])
    subRoot4 = build_tree([1])
    result4 = solution.isSubtree(root4, subRoot4)
    print(f"Result: {result4}, Expected: True")
    assert result4 == True, "Test 4 failed"

    # Test case 5: Single node trees, not matching
    print("\nTest 5: Single node trees, not matching")
    root5 = build_tree([1])
    subRoot5 = build_tree([2])
    result5 = solution.isSubtree(root5, subRoot5)
    print(f"Result: {result5}, Expected: False")
    assert result5 == False, "Test 5 failed"

    # Test case 6: subRoot at right child
    print("\nTest 6: subRoot at right child")
    root6 = build_tree([3, 4, 5, 1, 2])
    subRoot6 = build_tree([5])
    result6 = solution.isSubtree(root6, subRoot6)
    print(f"Result: {result6}, Expected: True")
    assert result6 == True, "Test 6 failed"

    # Test case 7: subRoot is leaf node
    print("\nTest 7: subRoot is leaf node")
    root7 = build_tree([3, 4, 5, 1, 2])
    subRoot7 = build_tree([2])
    result7 = solution.isSubtree(root7, subRoot7)
    print(f"Result: {result7}, Expected: True")
    assert result7 == True, "Test 7 failed"

    # Test case 8: Similar structure but different values
    print("\nTest 8: Similar structure but different values")
    root8 = build_tree([3, 4, 5, 1, 2])
    subRoot8 = build_tree([4, 1, 3])
    result8 = solution.isSubtree(root8, subRoot8)
    print(f"Result: {result8}, Expected: False")
    assert result8 == False, "Test 8 failed"

    # Test case 9: subRoot larger than root
    print("\nTest 9: subRoot larger than root")
    root9 = build_tree([1, 2])
    subRoot9 = build_tree([1, 2, 3])
    result9 = solution.isSubtree(root9, subRoot9)
    print(f"Result: {result9}, Expected: False")
    assert result9 == False, "Test 9 failed"

    # Test case 10: Complex tree, subRoot deep inside
    print("\nTest 10: Complex tree, subRoot deep inside")
    root10 = build_tree([1, 2, 3, 4, 5, 6, 7, 8, 9])
    subRoot10 = build_tree([4, 8, 9])
    result10 = solution.isSubtree(root10, subRoot10)
    print(f"Result: {result10}, Expected: True")
    assert result10 == True, "Test 10 failed"

    # Test case 11: Partial match (substring-like scenario)
    print("\nTest 11: Value appears but not as subtree")
    root11 = build_tree([12])
    subRoot11 = build_tree([2])
    result11 = solution.isSubtree(root11, subRoot11)
    print(f"Result: {result11}, Expected: False")
    assert result11 == False, "Test 11 failed"

    print("\n✅ All tests passed!")


if __name__ == "__main__":
    run_tests()
